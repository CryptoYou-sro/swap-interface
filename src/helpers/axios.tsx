import { useCallback, useEffect, useState } from 'react';
import { Buffer } from 'buffer';
import {
	BASE_URL,
	BINANCE_DEV_URL,
	BINANCE_SCRIPT,
	BIZ_ENTRY_KEY,
	BINANCE_EXCHANGE_INFO,
	BINANCE_PRICE_TICKER
} from './index';
import axios from 'axios';
import destinationNetworks from '../data/destinationNetworks.json';

export enum STATUS_ENUM {
	NONCE = 'NONCE',
	AUTH = 'AUTH',
	PASS = 'PASS'
}

type LocalStorageProps = {
	isKyced: boolean;
	refresh: string;
	access: string;
	account: string;
};

export const apiCall = {
	getNonce: 'nonce?address=',
	auth: 'auth',
	kycToken: 'kyc/token',
	kycStatus: 'kyc/status',
	refresh: 'refresh'
};

export const getMetamaskMessage = (nonce: string): string =>
	`0x${Buffer.from('Please sign this one time nonce: ' + nonce, 'utf8').toString('hex')}`;

export const loadBinanceKycScript = (cb?: any) => {
	const existingId = document.getElementById('binance-kcy-script');

	if (!existingId) {
		const binanceSdkScript = document.createElement('script');
		binanceSdkScript.src = BINANCE_SCRIPT;
		binanceSdkScript.id = 'binance-kcy-script';
		document.body.appendChild(binanceSdkScript);

		binanceSdkScript.onload = () => {
			if (cb) cb();
		};
	}
	if (existingId && cb) cb();
};

export const makeBinanceKycCall = (authToken: string) => {
	// @ts-ignore
	const binanceKyc = new BinanceKyc({
		authToken,
		bizEntityKey: BIZ_ENTRY_KEY,
		apiHost: BINANCE_DEV_URL,
		onMessage: ({ typeCode }: any) => {
			if (typeCode === '102') {
				binanceKyc.switchVisible(true);
			}
		}
	});
};

const getStorageValue = (key: string) => {
	const defaultValue = {
		access: '',
		refresh: '',
		account: '',
		isKyced: false
	} as LocalStorageProps;

	if (typeof window !== 'undefined') {
		const saved = localStorage.getItem(key);

		return saved ? (JSON.parse(saved) as LocalStorageProps) : defaultValue;
	}
};

export const useLocalStorage = (key: string) => {
	const [value, setValue] = useState(() => {
		return getStorageValue(key);
	});

	useEffect(() => {
		localStorage.setItem(key, JSON.stringify(value));
	}, [key, value]);

	return [value, setValue];
};

export const getAuthTokensFromNonce = async (account: string, library: any) => {
	try {
		const res = await axios.request({
			url: `${BASE_URL}${apiCall.getNonce}${account}`
		});
		try {
			const msg = getMetamaskMessage(res.data.nonce);
			const signature = await library?.send('personal_sign', [account, msg]);
			try {
				const tokenRes = await axios.request({
					url: `${BASE_URL}${apiCall.auth}`,
					method: 'POST',
					data: { address: account, signature }
				});

				return tokenRes.data as { access: string; is_kyced: boolean; refresh: string }; // TODO: define Type
			} catch (err: any) {
				throw new Error(err);
			}
		} catch (err: any) {
			throw new Error(err);
		}
	} catch (err: any) {
		throw new Error(err);
	}
};

export const useKyc = (
	authToken: string
): { loading: boolean; error: any; kycStatus: string; kycToken: string } => {
	const [loading, setLoading] = useState(false);
	const [error, setError] = useState(null);
	const [kycStatus, setKycStatus] = useState('');
	const [kycToken, setKycToken] = useState('');

	const fetchData = useCallback(
		async (authToken: string): Promise<void> => {
			try {
				setLoading(true);
				const statusRes = await axios.request({
					// TODO: check typing and if kycStatus is from right place
					url: `${BASE_URL}${apiCall.kycStatus}`,
					headers: {
						Authorization: `Bearer ${authToken}`
					}
				});
				setKycStatus(statusRes.data.statusInfo.kycStatus);
			} catch (err: any) {
				setError(err);
			} finally {
				setLoading(false);
			}

			try {
				setLoading(true);
				const tokenRes = await axios.request({
					url: `${BASE_URL}${apiCall.kycToken}`,
					headers: {
						Authorization: `Bearer ${authToken}`
					}
				});
				setKycToken(tokenRes.data.token);
			} catch (err: any) {
				setError(err);
			} finally {
				setLoading(false);
			}
		},
		[authToken]
	);

	useEffect(() => {
		void fetchData(authToken);
	}, [fetchData]);

	return { loading, error, kycStatus, kycToken };
};

export const useBinanceApi = () => {
	const [allPrices, setAllPrices] = useState<{ symbol: string; price: string }[]>([]);
	const [allFilteredPrices, setAllFilteredPrices] = useState<{ symbol: string; price: string }[]>(
		[]
	);
	const [allPairs, setAllPairs] = useState([]);
	const [allFilteredPairs, setAllFilteredPairs] = useState([]);

	const getExchangeInfo = async () => {
		try {
			const res = await axios.request({ url: BINANCE_EXCHANGE_INFO });
			setAllPairs(res.data.symbols);
		} catch (e: any) {
			throw new Error(e);
		}
	};

	const getTickerData = async () => {
		try {
			const res = await axios.request({ url: BINANCE_PRICE_TICKER });
			setAllPrices(res.data);
		} catch (e: any) {
			throw new Error(e);
		}
	};

	const uniqueTokens: string[] = Object.keys(destinationNetworks).reduce(
		(tokens: string[], network: string) => {
			// @ts-ignore
			const networkTokens = Object.keys(destinationNetworks?.[network]?.['tokens']);

			const allTokens = [...tokens, networkTokens];

			return [...new Set(allTokens.flat(1))];
		},
		['GLMR']
	);

	const isSymbol = (symbol: string): boolean => {
		let k = 0;
		for (let i = 0; i < uniqueTokens.length - 1; i++) {
			k++;
			for (let j = k; j < uniqueTokens.length; j++) {
				if (
					uniqueTokens[i] + uniqueTokens[j] === symbol ||
					uniqueTokens[j] + uniqueTokens[i] === symbol
				) {
					return true;
				}
			}
		}

		return false;
	};

	useEffect(() => {
		void getExchangeInfo();
		void getTickerData();
	}, []);

	useEffect(() => {
		if (allPairs) {
			const filteredPairs = allPairs.filter((pair: any) => isSymbol(pair.symbol));
			setAllFilteredPairs(filteredPairs);
		}

		if (allPrices) {
			const filteredPrices = allPrices.filter((price: any) => isSymbol(price.symbol));
			setAllFilteredPrices(filteredPrices);
		}
	}, [allPairs, allPrices]);

	return { allFilteredPairs, allFilteredPrices };
};
