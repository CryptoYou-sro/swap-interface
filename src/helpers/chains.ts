import { Chain } from '@wagmi/core';
 
export const moonbeam = {
    id: 1284,
    name: 'Moonbeam',
    network: 'moonbeam',
    nativeCurrency: {
        decimals: 18,
        name: 'GLMR',
        symbol: 'GLMR',
    },
    rpcUrls: {
        public: {
            http: ['https://rpc.api.moonbeam.network'],
            webSocket: ['wss://wss.api.moonbeam.network'],
        },
        default: {
            http: ['https://rpc.api.moonbeam.network'],
            webSocket: ['wss://wss.api.moonbeam.network'],
        },
    },
    blockExplorers: {
        default: {
            name: 'Moonscan',
            url: 'https://moonscan.io',
        },
        etherscan: {
            name: 'Moonscan',
            url: 'https://moonscan.io',
        },
    },
    contracts: {
        multicall3: {
            address: '0xcA11bde05977b3631167028862bE2a173976CA11',
            blockCreated: 609002,
        },
    },
    testnet: false
} as const satisfies Chain;

export const bsc = {
    id: 56,
    name: 'Binance Smart Chain',
    network: 'bsc',
    nativeCurrency: {
        decimals: 18,
        name: 'BNB',
        symbol: 'BNB',
    },
    rpcUrls: {
        public: {
            http: ['https://bsc.publicnode.com'],
        },
        default: {
            http: ['https://bsc.publicnode.com'],
        },
    },
    blockExplorers: {
        default: {
            name: 'BscScan',
            url: 'https://bscscan.com',
        },
        etherscan: {
            name: 'BscScan',
            url: 'https://bscscan.com',
        },
    },
    contracts: {
        multicall3: {
            address: '0xca11bde05977b3631167028862be2a173976ca11',
            blockCreated: 15921452,
        },
    },
    testnet: false
} as const satisfies Chain;
