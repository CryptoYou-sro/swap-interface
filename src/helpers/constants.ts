export const LOCAL_STORAGE_THEME = 'darkMode';
export const LOCAL_STORAGE_AUTH = 'auth';

export const INITIAL_STORAGE = {
	access: '',
	refresh: '',
	account: '',
	isKyced: false
};

export const MESSAGE = {
	ignore: 'ignore',
	failed: 'Request failed'
};

export const BIZ_ENTRY_KEY = 'BMYBIT';
export const BINANCE_PROD_URL = 'https://api.commonservice.io';
export const BINANCE_DEV_URL = 'https://dip-qacb.sdtaop.com';
export const BINANCE_SCRIPT =
	'https://static.saasexch.com/static/binance/static/kyc-ui/sdk/0.0.2/sdk.js';
export const BINANCE_PRICE_TICKER = 'https://www.binance.com/api/v3/ticker/price';
export const BINANCE_EXCHANGE_INFO = 'https://api.binance.com/api/v3/exchangeInfo';

export const BASE_URL = 'https://auth-app-aq3rv.ondigitalocean.app/';
// https://auth-app-aq3rv.ondigitalocean.app/account/withdraw/details?id=7ef91f76c79b4793b417e32fbffd73b8
export const MOONBEAM_URL = 'https://rpc.api.moonbeam.network';

export const routes = {
	getNonce: 'nonce?address=',
	auth: 'auth',
	kycToken: 'kyc/token',
	kycStatus: 'kyc/status',
	refresh: 'auth/refresh',
	transactionDetails: 'account/withdraw/details?id='
};

export const PROTOCOL_FEE = 0.002;
export const BINANCE_FEE = 0.002;
export const PROTOCOL_FEE_FACTOR = 1 / (1 - PROTOCOL_FEE);
export const START_TOKEN = 'GLMR';
export const FEE_CURRENCY = 'USDT';

export const ESTIMATED_NETWORK_TRANSACTION_GAS = 55_437;
export const SERVICE_ADDRESS = '0xeB56c1d19855cc0346f437028e6ad09C80128e02';

export const BLOCK_CONTRACT_NUMBER = 2_013_557;
export const WEI_TO_GLMR = 1 / 1000000000000000000;

export const CONTRACT_ADDRESSES = {
	1284: '0x6136F6c8FdC1c711892d217d0D5E95928dce6b20',
	1287: '0x2d249342F1F6D549CE8253782c8A2b9218c68fB2',
	31337: '0x0165878A594ca255338adfa4d48449f69242Eb8F'
};

export const BLOCKS_AMOUNT = 30;

export const makeId = (length: number) => {
	let result = '';
	const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
	const charactersLength = characters.length;
	for (let i = 0; i < length; i++) {
		result += characters.charAt(Math.floor(Math.random() * charactersLength));
	}

	return result;
};
