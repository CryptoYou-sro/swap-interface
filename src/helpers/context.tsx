import React, { createContext, ReactNode, useContext, useEffect, useReducer } from 'react';
import type { Theme } from '../styles';
import { darkTheme } from '../styles';
import type { ColorType } from '../components';

export enum VerificationEnum {
	ACCOUNT = 'SET_ACCOUNT_CONNECTED',
	NETWORK = 'SET_NETWORK_CONNECTED',
	USER = 'SET_USER_VERIFIED'
}

export enum ThemeEnum {
	THEME = 'SET_THEME'
}

export enum DestinationNetworkEnum {
	WALLET = 'SET_DESTINATION_WALLET',
	NETWORK = 'SET_DESTINATION_NETWORK',
	TOKEN = 'SET_DESTINATION_TOKEN',
	ADDRESS = 'SET_DESTINATION_ADDRESS',
	AMOUNT = 'SET_DESTINATION_AMOUNT',
	MEMO = 'SET_DESTINATION_MEMO'
}

export enum KycEnum {
	STATUS = 'SET_KYC_STATUS'
}

export enum KycStatusEnum {
	PROCESS = 'PROCESS',
	REVIEW = 'REVIEW',
	REFUSED = 'REFUSED',
	PASS = 'PASS',
	DISABLE = 'DISABLE',
	REJECT = 'REJECT'
}

export enum BasicStatusEnum {
	INITIAL = 'INITIAL',
	PROCESS = 'PROCESS',
	REVIEW = 'REVIEW',
	PASS = 'PASS',
	REJECT = 'REJECT',
	SKIP = 'SKIP'
}

export enum ButtonEnum {
	BUTTON = 'SET_BUTTON_STATUS'
}

type VerificationAction = {
	type: VerificationEnum;
	payload: boolean;
};

type KycAction = {
	type: KycEnum;
	payload: KycStatusEnum;
};

type ButtonAction = {
	type: ButtonEnum;
	payload: { color: string; text: string };
};

type ThemeAction = {
	type: ThemeEnum;
	payload: Theme;
};

type DestinationNetworkAction = {
	type: DestinationNetworkEnum;
	payload: string;
};

type Action =
	| VerificationAction
	| ButtonAction
	| KycAction
	| ThemeAction
	| DestinationNetworkAction;

type State = {
	isUserVerified: boolean;
	isAccountConnected: boolean;
	isNetworkConnected: boolean;
	kycStatus: KycStatusEnum;
	buttonStatus: { color: string; text: string };
	theme: Theme;
	destinationWallet: string;
	destinationNetwork: string;
	destinationToken: string;
	destinationAddress: string;
	destinationAmount: string;
	destinationMemo: string;
};

type ButtonStatus = {
	// TODO: refactor type
	CONNECT_WALLET: { color: ColorType; text: string };
	CHANGE_NETWORK: { color: ColorType; text: string };
	PASS_KYC: { color: ColorType; text: string };
	CHECK_KYC: { color: ColorType; text: string };
	// GET_NONCE: { color: ColorType; text: string };
};

export const buttonText = {
	CONNECT_WALLET: 'Connect Wallet',
	CHANGE_NETWORK: 'Change Network',
	PASS_KYC: 'Pass KYC',
	CHECK_KYC: 'Check KYC'
	// GET_NONCE: 'Get Nonce'
};

export const buttonType: ButtonStatus = {
	CONNECT_WALLET: { color: 'default', text: buttonText.CONNECT_WALLET },
	CHANGE_NETWORK: { color: 'error', text: buttonText.CHANGE_NETWORK },
	PASS_KYC: { color: 'warning', text: buttonText.PASS_KYC },
	CHECK_KYC: { color: 'success', text: buttonText.CHECK_KYC }
	// GET_NONCE: { color: 'default', text: buttonText.GET_NONCE }
};

const initialState: State = {
	isUserVerified: false,
	isAccountConnected: false,
	isNetworkConnected: false,
	kycStatus: KycStatusEnum.PROCESS,
	buttonStatus: buttonType.CONNECT_WALLET,
	theme: darkTheme,
	destinationWallet: 'Select Wallet',
	destinationNetwork: 'Select Network',
	destinationToken: 'Select Token',
	destinationAddress: '',
	destinationAmount: '',
	destinationMemo: ''
};

type Dispatch = (action: Action) => void;

const AuthContext = createContext<{ state: State; dispatch: Dispatch } | undefined>(undefined);

const authReducer = (state: State, action: Action): State => {
	switch (action.type) {
		case VerificationEnum.ACCOUNT:
			return { ...state, isAccountConnected: action.payload };
		case VerificationEnum.NETWORK:
			return { ...state, isNetworkConnected: action.payload };
		case KycEnum.STATUS:
			return { ...state, kycStatus: action.payload };
		case ButtonEnum.BUTTON:
			return { ...state, buttonStatus: action.payload };
		case VerificationEnum.USER:
			return { ...state, isUserVerified: action.payload };
		case ThemeEnum.THEME:
			return { ...state, theme: action.payload };
		case DestinationNetworkEnum.WALLET:
			return { ...state, destinationWallet: action.payload };
		case DestinationNetworkEnum.NETWORK:
			return { ...state, destinationNetwork: action.payload };
		case DestinationNetworkEnum.TOKEN:
			return { ...state, destinationToken: action.payload };
		case DestinationNetworkEnum.ADDRESS:
			return { ...state, destinationAddress: action.payload };
		case DestinationNetworkEnum.AMOUNT:
			return { ...state, destinationAmount: action.payload };
		case DestinationNetworkEnum.MEMO:
			return { ...state, destinationMemo: action.payload };
		default:
			return state;
	}
};

export const AuthProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
	const [state, dispatch] = useReducer(authReducer, initialState);
	const value = { state, dispatch };
	const { isAccountConnected, isNetworkConnected, kycStatus } = state;

	useEffect(() => {
		if (!isAccountConnected) {
			dispatch({
				type: ButtonEnum.BUTTON,
				payload: buttonType.CONNECT_WALLET
			});
		}

		if (!isNetworkConnected && isAccountConnected) {
			dispatch({
				type: ButtonEnum.BUTTON,
				payload: buttonType.CHANGE_NETWORK
			});
		}

		// if (
		// 	(kycStatus === KycStatusEnum.REJECT ||
		// 		kycStatus === KycStatusEnum.REFUSED ||
		// 		kycStatus === KycStatusEnum.PROCESS) &&
		// 	isNetworkConnected &&
		// 	isAccountConnected
		// ) {
		// 	dispatch({
		// 		type: ButtonEnum.BUTTON,
		// 		payload: buttonType.PASS_KYC
		// 	});
		// }

		// if (kycStatus === KycStatusEnum.DISABLE || kycStatus === KycStatusEnum.REVIEW) {
		// 	dispatch({
		// 		type: ButtonEnum.BUTTON,
		// 		payload: buttonType.CHECK_KYC
		// 	});
		// }

		if (kycStatus === KycStatusEnum.PASS && isNetworkConnected && isAccountConnected) {
			dispatch({ type: VerificationEnum.USER, payload: true });
		}

		if (kycStatus !== KycStatusEnum.PASS || !isNetworkConnected || !isAccountConnected) {
			dispatch({ type: VerificationEnum.USER, payload: false });
		}
	}, [isAccountConnected, isNetworkConnected, kycStatus]);

	return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export const useStore = () => {
	const context = useContext(AuthContext);
	if (context === undefined) {
		throw new Error('useStore must be used within a AuthProvider');
	}

	return context;
};
