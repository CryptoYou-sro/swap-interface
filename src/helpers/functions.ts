import type { Breakpoint, BreakpointOrNumber, Theme } from './../styles';
import { breakpoint } from './../styles';
import { useLayoutEffect, useState } from 'react';
import { format, utcToZonedTime } from 'date-fns-tz';

export const isLightTheme = (theme: Theme): boolean => theme.name === 'light';
const { timeZone } = Intl.DateTimeFormat().resolvedOptions();

export const isNetworkSelected = (network: string) =>
	network !== 'Select Network' && network !== ''; // TODO: refine both functions - nullish check - and create enum for "Select Network / Token"

export const isArrayType = (value: any) => typeof value === 'object' && Array.isArray(value);

export const isTokenSelected = (token: string) => token !== 'Select Token' && token !== '';

type BeautifyNumbers = {
	n: string | number;
	digits?: number;
};

const trimZeros = (res: string): string =>
	res.slice(-1) === '0' && res.slice(-2, -1) !== '.' ? trimZeros(res.slice(0, -1)) : res;

export const beautifyNumbers = ({ n, digits = 8 }: BeautifyNumbers): string => {
	let res = '';
	if (!n) return '';
	if (typeof n === 'number') {
		res = n.toFixed(digits);
	} else {
		res = (+n).toFixed(digits);
	}

	return trimZeros(res);
};

export const useWindowSize = () => {
	const [size, setSize] = useState([0, 0]);
	useLayoutEffect(() => {
		const updateSize = () => {
			setSize([window.innerWidth, window.innerHeight]);
		};
		window.addEventListener('resize', updateSize);
		updateSize();

		return () => window.removeEventListener('resize', updateSize);
	}, []);

	return size;
};

export const useBreakpoint = (size: BreakpointOrNumber) => {
	const [windowWidth, windowHeight] = useWindowSize();
	const isString = typeof size === typeof 'string';

	return {
		isBreakpointWidth: windowWidth < (isString ? breakpoint[size as Breakpoint] : size),
		isBreakpointHeight: windowHeight < (isString ? breakpoint[size as Breakpoint] : size)
	};
};

export const hexToRgbA = (hex: string, alpha = '1'): string => {
	const [r, g, b] = hex.match(/\w\w/g)!.map((x) => parseInt(x, 16));

	return `rgba(${r},${g},${b},${alpha})`;
};

export const isSwapRejected = (status: string, errorMessage: any) =>
	status === 'Exception' && errorMessage === 'user rejected transaction';

export const formatDate = (ts: number | undefined): string =>
	ts ? format(utcToZonedTime(new Date(ts * 1000), timeZone), 'dd/MM/yyyy HH:mm:ss') : 'n/a';
