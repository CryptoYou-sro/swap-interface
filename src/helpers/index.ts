export * from './functions';
export * from './context';
export * from './constants';
export * from './auth';
export * from './graph';
export * from './storage';
export * from './types';
