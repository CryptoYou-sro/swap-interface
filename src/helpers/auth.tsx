import { Buffer } from 'buffer';
import { BASE_URL, BINANCE_PROD_URL, BINANCE_SCRIPT, BIZ_ENTRY_KEY } from '../helpers';
import axios from 'axios';
import type { ApiAuthType } from '../helpers';

export enum STATUS_ENUM {
	NONCE = 'NONCE',
	AUTH = 'AUTH',
	PASS = 'PASS'
}

export const routes = {
	getNonce: 'nonce?address=',
	auth: 'auth',
	kycToken: 'kyc/token',
	kycStatus: 'kyc/status',
	refresh: 'auth/refresh'
};

export const getMetamaskMessage = (nonce: string): string =>
	`0x${Buffer.from('Please sign this one time nonce: ' + nonce, 'utf8').toString('hex')}`;

export const loadBinanceKycScript = (cb?: any) => {
	const existingId = document.getElementById('binance-kcy-script');

	if (!existingId) {
		const binanceSdkScript = document.createElement('script');
		binanceSdkScript.src = BINANCE_SCRIPT;
		binanceSdkScript.id = 'binance-kcy-script';
		document.body.appendChild(binanceSdkScript);

		binanceSdkScript.onload = () => {
			if (cb) cb();
		};
	}
	if (existingId && cb) cb();
};

export const makeBinanceKycCall = (authToken: string) => {
	// @ts-ignore
	const binanceKyc = new BinanceKyc({
		authToken,
		bizEntityKey: BIZ_ENTRY_KEY,
		apiHost: BINANCE_PROD_URL,
		onMessage: ({ typeCode }: any) => {
			if (typeCode === '102') {
				binanceKyc.switchVisible(true);
			}
		}
	});
};

export const getAuthTokensFromNonce = async (account: string, library: any) => {
	try {
		const res = await axios.request({
			url: `${BASE_URL}${routes.getNonce}${account}`
		});
		try {
			const msg = getMetamaskMessage(res.data.nonce);
			const signature = await library?.send('personal_sign', [account, msg]);
			try {
				const tokenRes = await axios.request({
					url: `${BASE_URL}${routes.auth}`,
					method: 'POST',
					data: { address: account, signature }
				});

				return tokenRes.data as ApiAuthType;
			} catch (err: any) {
				throw new Error(err);
			}
		} catch (err: any) {
			throw new Error(err);
		}
	} catch (err: any) {
		throw new Error(err);
	}
};
