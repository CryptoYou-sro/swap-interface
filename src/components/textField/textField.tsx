import { useState } from 'react';
import styled, { css } from 'styled-components';
import {
	DEFAULT_BORDER_RADIUS,
	HORIZONTAL_PADDING,
	fontSize,
	pxToRem,
	spacing,
	DEFAULT_TRANSIITON,
	DEFAULT_OUTLINE_OFFSET
} from '../../styles';
import type { ThemeProps } from '../../styles';
import { useStore } from '../../helpers';
import { Icon } from '../../components';

export type AlignProps = 'left' | 'right' | 'center';
export type TypeProps = 'text' | 'number' | 'search';
export type SizeProps = 'regular' | 'small';

type StyledProps = {
	align: AlignProps;
	error: boolean;
	type: TypeProps;
	size: SizeProps;
};

const TextFieldWrapper = styled.div`
	position: relative;
`;

const Input = styled.input(({ align, error, type, size }: StyledProps) => {
	const {
		state: { theme }
	} = useStore();

	const isTypeNumber = type === 'number';
	const isTypeSearch = type === 'search';
	const isSmall = size === 'small';

	return css`
		background: none;
		text-align: ${isTypeSearch ? 'left' : align};
		font-size: ${fontSize[16]};
		line-height: ${fontSize[20]};
		padding: ${isTypeSearch
			? `${spacing[isSmall ? 8 : 14]} ${spacing[14]} ${spacing[isSmall ? 8 : 14]} ${spacing[42]}`
			: `${spacing[isSmall ? 12 : 18]} ${spacing[HORIZONTAL_PADDING]}`};
		color: ${theme.font.default};
		border: 1px solid ${error && isTypeNumber ? theme.button.error : theme.border.default};
		border-radius: ${DEFAULT_BORDER_RADIUS};
		cursor: pointer;
		transition: ${DEFAULT_TRANSIITON};
		width: ${isTypeSearch ? '100%' : `calc(100% - ${pxToRem(HORIZONTAL_PADDING * 2 + 2)})`};
		outline: 1px solid transparent;

		&:hover,
		&:active {
			border-color: ${error && isTypeNumber ? theme.button.error : theme.font.secondary};
			outline: none;
		}

		&:focus-visible {
			outline-offset: ${DEFAULT_OUTLINE_OFFSET};
			outline: 1px solid ${error && isTypeNumber ? theme.button.error : theme.border.secondary};
		}

		&-webkit-outer-spin-button,
		&::-webkit-inner-spin-button {
			-webkit-appearance: none;
			margin: 0;
		}

		&[type='search']::-webkit-search-cancel-button {
			// TODO: do we not want a cancel button within the search?
			-webkit-appearance: none;
			appearance: none;
		}

		&[type='number'] {
			-moz-appearance: textfield;
		}
	`;
});

const Message = styled.div`
	display: flex;
	justify-content: space-between;
`;

const Description = styled.div`
	color: ${(props: ThemeProps) => props.theme.font.secondary};
	margin: ${spacing[4]} 0;
`;

const Error = styled.div`
	margin: ${spacing[4]} 0;
	color: ${(props: ThemeProps) => props.theme.button.error};
`;

type Props = {
	placeholder?: string;
	disabled?: boolean;
	type?: TypeProps;
	value: string;
	description?: string;
	error?: boolean;
	size?: SizeProps;
	onChange?: (e?: any) => void;
	align?: AlignProps;
};

export const TextField = ({
	placeholder,
	disabled = false,
	type = 'text',
	value,
	onChange,
	description,
	error,
	size = 'regular',
	align = 'center'
}: Props) => {
	const {
		state: { theme }
	} = useStore();
	const [isActive, setIsActive] = useState(false);
	const isTypeSearch = type === 'search';

	const textField = (
		<>
			{/* @ts-ignore */}
			<Input
				placeholder={placeholder}
				disabled={disabled}
				onChange={onChange}
				align={align}
				value={value}
				type={type}
				min="0"
				size={size}
				error={error}
				onBlur={() => setIsActive(true)}
				onFocus={() => setIsActive(false)}
			/>
			{isTypeSearch && (
				<Icon
					icon="search"
					size={20}
					style={{
						position: 'absolute',
						transform: ' translate(-50%, -50%)',
						top: '50%',
						left: 20
					}}
				/>
			)}
			{(error || description) && type === 'text' && (
				<Message>
					{description && <Description theme={theme}>{description}</Description>}
					{error && isActive && <Error theme={theme}>Invalid input</Error>}
				</Message>
			)}
		</>
	);

	return isTypeSearch ? <TextFieldWrapper>{textField}</TextFieldWrapper> : textField;
};
