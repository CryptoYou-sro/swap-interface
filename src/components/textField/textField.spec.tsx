import { TextField } from './textField';
import { render } from '@testing-library/react';
import { AuthProvider } from '../../helpers';

describe('TextField', () => {
	it('should match snapshot', () => {
		const { getByPlaceholderText } = render(
			<AuthProvider>
				<TextField
					disabled
					value="Test Value"
					placeholder="placeholder"
					type="number"
					description={'Text Field Component'} />
			</AuthProvider>
		);
		const textField = getByPlaceholderText(/placeholder/i);
		expect(textField).toMatchSnapshot();
	});

	it('should match style', () => {
		const { getByPlaceholderText } = render(
			<AuthProvider>
				<TextField
					disabled
					value="Test Value"
					placeholder="placeholder"
					type="number"
					description={'Text Field Component'} />
			</AuthProvider>
		);
		const textField = getByPlaceholderText(/placeholder/i);
		expect(textField).toHaveStyle('border-radius: 0.375rem; font-size: 1rem; line-height: 1.25rem; padding: 1.125rem 0;');
	});

	it('should set props correctly', () => {
		const { getByPlaceholderText } = render(
			<AuthProvider>
				<TextField
					disabled
					value="value"
					placeholder="placeholder"
					type="number"
				/>
			</AuthProvider>
		);
		const textField = getByPlaceholderText(/placeholder/i) as HTMLInputElement;
		expect(textField.placeholder).toBe('placeholder');
		expect(textField.readOnly).toBe(true);
		expect(textField.type).toBe('number');
		expect(textField.lang).toBe('en');
	});
});
