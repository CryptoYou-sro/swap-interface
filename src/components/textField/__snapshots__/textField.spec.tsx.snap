// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField should match snapshot for value disabled: false, type: number, size: small, align: left, error: true and description: description 1`] = `
<input
  class="sc-gJqSRm hxNZXQ"
  placeholder="placeholder"
  type="number"
  value="Test value"
/>
`;

exports[`TextField should match snapshot for value disabled: false, type: text, size: small, align: left, error: false and description: description 1`] = `
<input
  class="sc-gJqSRm khUawJ"
  placeholder="placeholder"
  type="text"
  value="Test value"
/>
`;

exports[`TextField should match snapshot for value disabled: true, type: number, size: regular, align: right, error: true and description: description 1`] = `
<input
  class="sc-gJqSRm gJsZhK"
  disabled=""
  placeholder="placeholder"
  type="number"
  value="Test value"
/>
`;

exports[`TextField should match snapshot for value disabled: true, type: search, size: small, align: center, error: false and description:  1`] = `
<input
  class="sc-gJqSRm kAeSHJ"
  disabled=""
  placeholder="placeholder"
  type="search"
  value="Test value"
/>
`;
